google/cloud/storage/client_test.cc

```diff
@@ -221,6 +221,38 @@ TEST_F(ClientTest, LoggingDecoratorsRestClient) {
                           "RetryClient"));
 }
 
+/// @test Verify the constructor creates the right set of RawClient decorations.
+TEST_F(ClientTest, FullStack) {
+  // Reset environment variables. We want the `Options` parameter to control the
+  // structure of the stack.
+  ScopedEnvironment logging("CLOUD_STORAGE_ENABLE_TRACING", absl::nullopt);
+  ScopedEnvironment legacy("GOOGLE_CLOUD_CPP_STORAGE_USE_LEGACY_HTTP",
+                           absl::nullopt);
+
+  // Create a client. Use the anonymous credentials because on the CI
+  // environment there may not be other credentials configured.
+  auto options = Options{}
+                     .set<UnifiedCredentialsOption>(MakeInsecureCredentials())
+                     .set<internal::UseRestClientOption>(true)
+                     .set<TracingComponentsOption>({"raw-client"});
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+  options.set<experimental::OpenTelemetryTracingOption>(true);
+#endif  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+  auto tested = Client(options);
+  auto const impl = ClientImplDetails::GetRawClient(tested);
+  ASSERT_THAT(impl, NotNull());
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+  EXPECT_THAT(impl->InspectStackStructure(),
+              ElementsAre("RestClient", "LoggingClient", "GenericStubAdapter",
+                          "RetryClient", "TracingClient"));
+#else
+  EXPECT_THAT(impl->InspectStackStructure(),
+              ElementsAre("RestClient", "LoggingClient", "GenericStubAdapter",
+                          "RetryClient"));
+#endif  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
```

Comment:
I think the `else` branch of this is uninteresting, because it is already tested above:

https://github.com/googleapis/google-cloud-cpp/blob/2c77f95a223441f3e20006de2e5cc1cb20d013fd/google/cloud/storage/client_test.cc#L219-L221

I would write the code like:

```cc
#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
using ::google::cloud::testing_util::DisableTracing;
using ::google::cloud::testing_util::EnableTracing;

TEST_F(ClientTest, TracingEnabled) {
  // blah blah blah
  auto options = EnableTracing(...);
  // blah blah blah
  EXPECT(..., "TracingClient", ...);
}

TEST_F(ClientTest, TracingDisabled) {
  // blah blah blah
  auto options = DisableTracing(...);
  // blah blah blah
  EXPECT(..., Not(Contains("TracingClient")));
}
#endif // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
```

Reply:
I think we want both an enabled and disabled test to catch things like #12098